### PR TITLE
Added None check for firewall rule attributes.

### PR DESCRIPTION
**Pull request by wyscan**
##### ISSUE TYPE
 - Bugfix Pull Request

##### SUMMARY
Fix for Issue #20481

### CODE REVIEW

File: lib/ansible/modules/cloud/google/gce_net.py [hunk 1]
```diff
@@ -359,27 +359,27 @@ def main():
 
                 # If these attributes are lists, we sort them first, then compare.
                 # Otherwise, we update if they differ.
-                if fw.source_ranges != src_range:
+                if fw.source_ranges != src_range and (fw.source_ranges or src_range):
```
Comment (supertom) [inline lib/ansible/modules/cloud/google/gce_net.py]: I'm not sure if this check is actually adding value.  Seems to me that if they are both None, it won't enter the block anyway due to the equality check on the left.

A quick test shows me that the playbook works without it.  Is it OK to remove these from this and the other if-blocks?

File: lib/ansible/modules/cloud/google/gce_net.py [hunk 2]
```diff
@@ -359,27 +359,27 @@ def main():
...
-                        if sorted(fw.source_ranges) != sorted(src_range):
+                        if fw.source_ranges is None or sorted(fw.source_ranges) != sorted(src_range):
                             fw.source_ranges = src_range
                             fw_changed = True
                     else:
                         fw.source_ranges = src_range
                         fw_changed = True
 
-                if fw.source_tags != src_tags:
+                if fw.source_tags != src_tags and (fw.source_tags or src_tags):
```
Comment (supertom) [inline lib/ansible/modules/cloud/google/gce_net.py]: Same comment as above.

File: lib/ansible/modules/cloud/google/gce_net.py [hunk 3]
```diff
@@ -359,27 +359,27 @@ def main():
...
-                        if sorted(fw.source_tags) != sorted(src_tags):
+                        if fw.source_tags is None or sorted(fw.source_tags) != sorted(src_tags):
                             fw.source_tags = src_tags
                             fw_changed = True
                     else:
                         fw.source_tags = src_tags
                         fw_changed = True
 
-                if fw.target_tags != target_tags:
+                if fw.target_tags != target_tags and (fw.target_tags or target_tags):
```
Comment (supertom) [inline lib/ansible/modules/cloud/google/gce_net.py]: Same comment as above.

**Review by supertom — CHANGES_REQUESTED** [review]
Please have a look at the review comments.

**Comment by wyscan** [comment]
I finally figured out what is going on. The issue is that rules are being updated unnecessarily when an attribute in the role is an empty list and the corresponding attribute in the cloud project is also empty. GCE returns the attribute as null. This python snippet may help explain it better:
```
#!/usr/bin/python

test1 = []
test2 = None

print test1 == True
print test2 == True
print test1 == test2
```
All of the comparisons are false. You should also be able to see the effect via this test role:
```
- name: Create Firewall Rule
  gce_net:
    name: default
    credentials_file: "{{ gce_credentials_file }}"
    service_account_email: "{{ gce_service_account }}"
    project_id: "{{ gce_project_id }}"
    fwname: "{{ fwname }}"
    allowed: tcp:80
    state: "{{ state }}"
    src_range: []
    src_tags: "foo,bar,baz"
    target_tags: ["testing"]
  register: result

- name: "assert fw created"
  assert:
    that:
      - 'result.changed'
      - 'result.state == "{{ state }}"'

- name: Update Firewall Rule
  gce_net:
    name: default
    credentials_file: "{{ gce_credentials_file }}"
    service_account_email: "{{ gce_service_account }}"
    project_id: "{{ gce_project_id }}"
    fwname: "{{ fwname }}"
    allowed: tcp:80
    state: "{{ state }}"
    src_range: []
    src_tags: "foo,bar,baz"
    target_tags: ["testing"]
  register: result

-  name: "assert fw unchanged"
   assert:
     that:
       - 'result.changed==false'
```

I'm re-adding the check which I think represents the safest approach in terms of ensuring idempotency, but is admittedly not related to the initial issue...

**Comment by supertom** [comment]
Hi @wyscan, thanks for running this down and your example.  Given that it was difficult to find and we don't want other to suffer, I think we should be explicit about setting those values to what we expect.  I feel the None and extra condition by themselves hide the issue.

how about something like:

```
# source_ranges might not be set in the project; cast it to an empty list
fw.source_ranges = fw.source_ranges or []
```
at the the top of the block?  I haven't totally thought this out, but I'd the code to better document what is happening.  Thoughts?

**Comment by wyscan** [comment]
I agree, that would be a good idea. I believe the latest changes capture that and they pass all of the existing scenarios I was testing against.

**Comment by supertom** [comment]
Thanks @wyscan, I re-tested and LGTM!  @ryansb OK by me to merge.

**Comment by ryansb** [comment]
bot_status

**Comment by ryansb** [comment]
!needs_revision

**Review by ryansb — APPROVED** [review]

**Pull request merged**